src/rpc/Errors.h

```diff
@@ -0,0 +1,206 @@
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace RPC {
+
+/**
+ * @brief Custom clio RPC Errors.
+ */
+enum class ClioError {
+    rpcMALFORMED_CURRENCY = 5000,
+};
+
+/**
+ * @brief Holds info about a particular @ref ClioError.
+ */
+struct ClioErrorInfo
+{
+    ClioError const code;
+    std::string_view const error;
+    std::string_view const message;
+};
+
+/**
+ * @brief Clio uses compatible Rippled error codes for most RPC errors.
+ */
+using Error = ripple::error_code_i;
+
+/**
+ * @brief Clio operates on a combination of Rippled and Custom Clio error codes.
+ *
+ * @see Error For rippled error codes
+ * @see ClioError For custom clio error codes
+ */
+using CombinedError = std::variant<Error, ClioError>;
+
+/**
+ * @brief A status returned from any RPC handler.
+ */
+struct Status
+{
+    CombinedError code = Error::rpcSUCCESS;
+    std::string error = "";
+    std::string message = "";
+
+    Status() = default;
+    /* implicit */ Status(CombinedError code) : code(code){};
+
+    // HACK. Some rippled handlers explicitly specify errors.
+    // This means that we have to be able to duplicate this
+    // functionality.
+    explicit Status(std::string const& message)
+        : code(ripple::rpcUNKNOWN), message(message)
+    {
+    }
+
+    Status(CombinedError code, std::string message)
+        : code(code), message(message)
+    {
+    }
+
+    Status(CombinedError code, std::string error, std::string message)
+        : code(code), error(error), message(message)
+    {
+    }
+
+    /**
+     * @brief Returns true if the Status is *not* OK.
+     */
+    operator bool() const
+    {
+        if (auto err = std::get_if<Error>(&code))
+            return *err != Error::rpcSUCCESS;
+        return true;
+    }
+};
+
+/**
+ * @brief Warning codes that can be returned by clio.
+ */
+enum WarningCode {
+    warnUNKNOWN = -1,
+    warnRPC_CLIO = 2001,
+    warnRPC_OUTDATED = 2002,
+    warnRPC_RATE_LIMIT = 2003
+};
+
+/**
+ * @brief Holds information about a clio warning.
+ */
+struct WarningInfo
+{
+    constexpr WarningInfo() = default;
+    constexpr WarningInfo(WarningCode code, char const* message)
+        : code(code), message(message)
+    {
+    }
+
+    WarningCode code = warnUNKNOWN;
+    std::string_view const message = "unknown warning";
+};
+
+/**
+ * @brief Invalid parameters error.
+ */
+class InvalidParamsError : public std::exception
+{
+    std::string msg;
+
+public:
+    explicit InvalidParamsError(std::string const& msg) : msg(msg)
+    {
+    }
+
+    const char*
+    what() const throw() override
+    {
+        return msg.c_str();
+    }
+};
+
+/**
+ * @brief Account not found error.
+ */
+class AccountNotFoundError : public std::exception
+{
+    std::string account;
+
+public:
+    explicit AccountNotFoundError(std::string const& acct) : account(acct)
+    {
+    }
+    const char*
+    what() const throw() override
+    {
+        return account.c_str();
+    }
+};
```

Comment:
Do we actually use these? I know they were here before.

Reply:
Yes, seems like we still do use them. We could possibly get rid of them eventually but i'd keep for now to make the changes in this pr minimal (already quite a lot).